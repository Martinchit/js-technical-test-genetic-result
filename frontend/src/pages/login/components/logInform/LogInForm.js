import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Button, Form, Message } from 'semantic-ui-react';
import { FlexBox } from '../../../../core/components/atoms/FlexBox';
import { logInAction, reset } from '../../../../redux/auth/authSlice';
import {
  emailPattern,
  passwordPattern,
} from '../../../../core/lib/utils/regexPatterns';

export const LogInForm = () => {
  const dispatch = useDispatch();
  const logInProcessing = useSelector((state) => state.auth.logInProcessing);
  const logInError = useSelector((state) => state.auth.logInError);
  const [email, setEmail] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [emailInvalidMessage] = React.useState(
    'Please make sure email is in correct pattern'
  );
  const [passwordInvalidMessage] = React.useState(
    'Please make sure password contains 8 alphanumeric characters'
  );

  const buttonContent = React.useMemo(() => {
    return logInProcessing ? 'Loading' : 'Log In';
  }, [logInProcessing]);

  const [isEmailInvalid, isPasswordInvalid] = React.useMemo(() => {
    const isEmailInvalid = !email || !emailPattern.test(email);
    const isPasswordInvalid = !password || !passwordPattern.test(password);
    return [isEmailInvalid, isPasswordInvalid];
  }, [email, password]);

  const updateEmail = React.useCallback(
    ({ target }) => {
      dispatch(reset());
      setEmail(target.value);
    },
    [dispatch]
  );

  const updatePassword = React.useCallback(
    ({ target }) => {
      dispatch(reset());
      setPassword(target.value);
    },
    [dispatch]
  );

  const logInFormOnSubmit = React.useCallback(() => {
    dispatch(logInAction(email, password));
  }, [dispatch, email, password]);

  return (
    <FlexBox flexDirection="column" height="auto">
      <Form
        error={logInError}
        warning={isEmailInvalid || isPasswordInvalid ? true : null}
      >
        <Form.Field>
          <Form.Input
            fluid
            label="Email"
            placeholder="Email"
            type="email"
            value={email}
            onChange={updateEmail}
          />
        </Form.Field>
        <Form.Field>
          <Form.Input
            fluid
            label="Email"
            placeholder="Password"
            type="password"
            value={password}
            onChange={updatePassword}
          />
        </Form.Field>
        <Message error content={logInError} />
        <Message
          warning
          content={
            isEmailInvalid
              ? emailInvalidMessage
              : isPasswordInvalid
              ? passwordInvalidMessage
              : null
          }
        />
        <Button
          color="green"
          disabled={isPasswordInvalid || isEmailInvalid}
          type="submit"
          onClick={logInFormOnSubmit}
        >
          {buttonContent}
        </Button>
      </Form>
    </FlexBox>
  );
};
