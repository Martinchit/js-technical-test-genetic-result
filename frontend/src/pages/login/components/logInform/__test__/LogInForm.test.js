import React from "react";
import { shallow } from 'enzyme';
import * as redux from 'react-redux';
import 'jest-styled-components';
import { Button, Form } from 'semantic-ui-react';

import { LogInForm } from "../LogInForm";

describe("LogInForm", () => {
  let spyOnUseSelector;
  let spyOnUseDispatch;
  let mockDispatch;
  const mockState = {
    auth: {
      logInProcessing: false,
      logInError: null
    }
  };

  beforeEach(() => {
    spyOnUseSelector = jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(mockState));
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });
  it("renders correctly", () => {
    const wrapper = shallow(<LogInForm />)
    expect(wrapper).toMatchSnapshot()
  })

  it("handles changes correctly", () => {
    const emailTestValue = 'email';
    const passwordTestValue = '123'
    const validEmailTestValue = 'test@test.com'
    const validPasswordTestValue = 'password'
    const wrapper = shallow(<LogInForm />)
    expect(wrapper).toMatchSnapshot()

    wrapper.find(Form.Input).at(0).simulate('change', { target: { value: emailTestValue }})
    expect(wrapper.find(Form.Input).at(0).props().value).toBe(emailTestValue)
    wrapper.find(Form.Input).at(1).simulate('change', { target: { value: passwordTestValue }})
    expect(wrapper.find(Form.Input).at(1).props().value).toBe(passwordTestValue)
    expect(wrapper.find(Button).at(0).props().disabled).toBe(true)
  
    wrapper.find(Form.Input).at(0).simulate('change', { target: { value: validEmailTestValue }})
    expect(wrapper.find(Form.Input).at(0).props().value).toBe(validEmailTestValue)
    expect(wrapper.find(Button).at(0).props().disabled).toBe(true)
  
    wrapper.find(Form.Input).at(1).simulate('change', { target: { value: validPasswordTestValue }})
    expect(wrapper.find(Form.Input).at(1).props().value).toBe(validPasswordTestValue)
    expect(wrapper.find(Button).at(0).props().disabled).toBe(false)

    wrapper.find(Button).at(0).simulate('click');
    expect(mockDispatch.mock.calls[0][0]).toEqual({ type: 'auth/reset', payload: undefined });
  })
})