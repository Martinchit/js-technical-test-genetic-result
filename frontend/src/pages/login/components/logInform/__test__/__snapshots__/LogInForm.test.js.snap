// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogInForm handles changes correctly 1`] = `ShallowWrapper {}`;

exports[`LogInForm renders correctly 1`] = `ShallowWrapper {}`;
